File: 70-tools/gftd/gftd/xrpc.go
// xrpc.go — gftd xrpc: invoke any XRPC endpoint on an App or PDS.
//
// Designed for use by Claude Code chat agents to trigger commands on deployed apps.
//
// Examples:
//
//	# Seed all platforms on media-gamers (nanoid a7m8oocs)
//	gftd xrpc ai.gftd.apps.media_gamers.catalog.seedAll -d '{"step":"platforms"}' --app a7m8oocs
//
//	# Seed games batch 0-9
//	gftd xrpc ai.gftd.apps.media_gamers.catalog.seedGames -d '{"offset":0,"limit":10}' --app a7m8oocs
//
//	# Generate guides for a specific game
//	gftd xrpc ai.gftd.apps.media_gamers.catalog.generateAll -d '{"slug":"elden-ring"}' --app a7m8oocs
//
//	# Query (GET) — no body
//	gftd xrpc ai.gftd.apps.media_gamers.catalog.getGame --app a7m8oocs
//
//	# Route through PDS (default when --app is omitted)
//	gftd xrpc com.atproto.server.describeServer
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
)

func runXRPC(args []string) error {
	fs := flag.NewFlagSet("xrpc", flag.ContinueOnError)
	data := fs.String("d", "", "JSON body for POST requests")
	appNanoid := fs.String("app", "", "App nanoid (routes to https://{nanoid}.etzhayyim.com/xrpc/{nsid})")
	baseURL := fs.String("url", "", "Override base URL (e.g. https://a7m8oocs.etzhayyim.com)")
	method := fs.String("X", "", "HTTP method: GET or POST (inferred from -d if omitted)")
	jsonOut := fs.Bool("json", false, "Pretty-print JSON response (default: raw stdout)")
	verbose := fs.Bool("v", false, "Verbose: print URL and status code to stderr")

	// Go's flag package stops at the first non-flag argument (the NSID), so flags
	// that follow the NSID would be silently ignored. Pre-pass: extract the NSID
	// (first arg that looks like an NSID — contains dots, no leading dash) and move
	// it to the end so the flag parser sees all flags first.
	nsid, flagArgs := extractNSID(args)

	if err := fs.Parse(flagArgs); err != nil {
		if err == flag.ErrHelp {
			return nil
		}
		return err
	}

	if nsid == "" {
		return fmt.Errorf("usage: gftd xrpc <nsid> [-d <json>] [--app <nanoid>] [--url <base>]\n\nExamples:\n  gftd xrpc ai.gftd.apps.media_gamers.catalog.seedAll -d '{\"step\":\"platforms\"}' --app a7m8oocs\n  gftd xrpc ai.gftd.apps.media_gamers.catalog.seedGames -d '{\"offset\":0,\"limit\":10}' --app a7m8oocs\n  gftd xrpc ai.gftd.apps.media_gamers.catalog.generateAll -d '{\"slug\":\"elden-ring\"}' --app a7m8oocs\n  gftd xrpc com.atproto.server.describeServer")
	}

	// Resolve base URL
	base := resolveXRPCBaseURL(*baseURL, *appNanoid, nsid)

	endpoint := strings.TrimRight(base, "/") + "/xrpc/" + nsid

	// Infer HTTP method
	httpMethod := strings.ToUpper(*method)
	if httpMethod == "" {
		if *data != "" {
			httpMethod = "POST"
		} else {
			httpMethod = "GET"
		}
	}

	// Build request
	var body io.Reader
	if *data != "" {
		body = bytes.NewBufferString(*data)
	}

	req, err := http.NewRequest(httpMethod, endpoint, body)
	if err != nil {
		return fmt.Errorf("build request: %w", err)
	}

	if *data != "" {
		req.Header.Set("Content-Type", "application/json")
	}
	req.Header.Set("Accept", "application/json")
	setAuthHeaders(req)

	if *verbose {
		fmt.Fprintf(os.Stderr, "[xrpc] %s %s\n", httpMethod, endpoint)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("read response: %w", err)
	}

	if *verbose {
		fmt.Fprintf(os.Stderr, "[xrpc] status: %s\n", resp.Status)
	}

	// Print response
	if *jsonOut && json.Valid(respBody) {
		var out any
		if err := json.Unmarshal(respBody, &out); err == nil {
			enc := json.NewEncoder(os.Stdout)
			enc.SetIndent("", "  ")
			_ = enc.Encode(out)
			return nil
		}
	}

	os.Stdout.Write(respBody)
	if len(respBody) > 0 && respBody[len(respBody)-1] != '\n' {
		fmt.Println()
	}

	// Non-2xx = error
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("XRPC %s: HTTP %s", nsid, resp.Status)
	}
	return nil
}

// extractNSID separates the NSID positional argument from flag arguments.
// The NSID is identified as the first arg that doesn't start with '-' and
// contains at least one dot (NSID always looks like com.atproto.* or ai.gftd.*).
// All other args are returned as flagArgs for flag.Parse().
func extractNSID(args []string) (nsid string, flagArgs []string) {
	for i := 0; i < len(args); i++ {
		a := args[i]
		if !strings.HasPrefix(a, "-") && strings.Contains(a, ".") && nsid == "" {
			nsid = a
			continue
		}
		flagArgs = append(flagArgs, a)
		// If this is a flag that takes a value (single-char or known multi-char),
		// consume the next arg as its value so we don't mis-classify it.
		if strings.HasPrefix(a, "-") && !strings.Contains(a, "=") {
			stripped := strings.TrimLeft(a, "-")
			if stripped == "d" || stripped == "X" || stripped == "app" || stripped == "url" {
				if i+1 < len(args) {
					i++
					flagArgs = append(flagArgs, args[i])
				}
			}
		}
	}
	return nsid, flagArgs
}

// resolveXRPCBaseURL resolves the base URL for an XRPC call.
//
// Priority:
//  1. --url flag (explicit override)
//  2. --app <nanoid> → https://{nanoid}.etzhayyim.com
//  3. Infer from NSID prefix (ai.gftd.apps.{app_slug}.*) via known nanoid map
//  4. Fall back to PDS (https://atproto.etzhayyim.com)
func resolveXRPCBaseURL(flagURL, appNanoid, nsid string) string {
	if flagURL != "" {
		return flagURL
	}
	if appNanoid != "" {
		return "https://" + appNanoid + ".etzhayyim.com"
	}
	// Infer from NSID: ai.gftd.apps.{slug}.* → look up in knownApps
	if b := inferBaseFromNSID(nsid); b != "" {
		return b
	}
	return resolvePDSBaseURL()
}

// knownApps maps app slug (from NSID 4th segment) to nanoid.
// Extend as new apps are registered.
var knownApps = map[string]string{
	"media_gamers":  "a7m8oocs",
	"media_anime":   "animegr01",
	"autorace":      "q7v8yed1k",
	"keirin":        "k31r1njp",
	"kyotei":        "qv8yed1k",
	"handotai":      "dtyy44cr",
	"hanrei":        "h4nr31jp",
	"kakaku":        "k4k4kux1",
	"gtin":          "gt1n4k7m",
}

func inferBaseFromNSID(nsid string) string {
	// NSID pattern: ai.gftd.apps.{slug}.{method}
	parts := strings.Split(nsid, ".")
	if len(parts) >= 4 && parts[0] == "ai" && parts[1] == "gftd" && parts[2] == "apps" {
		slug := parts[3]
		if nanoid, ok := knownApps[slug]; ok {
			return "https://" + nanoid + ".etzhayyim.com"
		}
	}
	return ""
}

File: 70-tools/etzhayyim-py/src/etzhayyim/xrpc.py
"""etzhayyim xrpc — port of xrpc.go (ADR-2605151500).

Invoke any XRPC endpoint on an App or PDS.
Auto-routes com.etzhayyim.apps.{slug}.* NSIDs to the correct nanoid worker.
Scoped-JWT auto-wrap: uses com.atproto.server.getServiceAuth for each NSID call.
"""

import json as _json_mod
import sys

import click
import httpx

from .auth import auth_headers, scoped_auth_headers, resolve_pds

_KNOWN_APPS: dict[str, str] = {
    "media_gamers": "a7m8oocs",
    "media_anime": "animegr01",
    "autorace": "q7v8yed1k",
    "keirin": "k31r1njp",
    "kyotei": "qv8yed1k",
    "handotai": "dtyy44cr",
    "hanrei": "h4nr31jp",
    "kakaku": "k4k4kux1",
    "gtin": "gt1n4k7m",
}

_APP_HOST_TEMPLATE = "https://{nanoid}.com.etzhayyim.com"


def _resolve_base(nsid: str, app: str | None, url: str | None) -> str:
    if url:
        return url.rstrip("/")
    if app:
        return _APP_HOST_TEMPLATE.format(nanoid=app)
    # NSID inference: com.etzhayyim.apps.<slug>.*
    parts = nsid.split(".")
    if len(parts) >= 4 and parts[:3] == ["com", "etzhayyim", "apps"]:
        slug = parts[3]
        if nanoid := _KNOWN_APPS.get(slug):
            return _APP_HOST_TEMPLATE.format(nanoid=nanoid)
    return resolve_pds()


@click.command("xrpc")
@click.argument("nsid")
@click.option("-d", "--data", "data_json", default=None, help="JSON body (POST) or query params (GET).")
@click.option("--app", default=None, help="Nanoid of target App Worker (overrides NSID inference).")
@click.option("--url", "base_url", default=None, help="Full base URL (overrides --app and inference).")
@click.option("-X", "--method", "http_method", default=None, type=click.Choice(["GET", "POST"], case_sensitive=False),
              help="HTTP method. Default: POST if -d given, else GET.")
@click.option("--json", "output_json", is_flag=True, help="Pretty-print JSON response.")
@click.option("-v", "--verbose", is_flag=True)
def xrpc(nsid: str, data_json: str | None, app: str | None, base_url: str | None,
         http_method: str | None, output_json: bool, verbose: bool) -> None:
    """Invoke an XRPC endpoint. NSID is the method name (e.g. com.atproto.server.describeServer)."""
    base = _resolve_base(nsid, app, base_url)
    url = f"{base}/xrpc/{nsid}"

    payload = None
    if data_json:
        try:
            payload = _json_mod.loads(data_json)
        except _json_mod.JSONDecodeError as exc:
            raise click.ClickException(f"Invalid JSON for -d: {exc}") from exc

    method = (http_method or ("POST" if payload is not None else "GET")).upper()
    headers = scoped_auth_headers(nsid)

    if verbose:
        click.echo(f"{method} {url}", err=True)
        if payload:
            click.echo(f"  body: {payload}", err=True)

    try:
        if method == "POST":
            resp = httpx.post(url, json=payload, headers=headers, timeout=30)
        else:
            params = payload if isinstance(payload, dict) else None
            resp = httpx.get(url, params=params, headers=headers, timeout=30)
    except httpx.RequestError as exc:
        raise click.ClickException(f"Request failed: {exc}") from exc

    if resp.status_code < 200 or resp.status_code >= 300:
        raise click.ClickException(f"XRPC {resp.status_code}: {resp.text}")

    body = resp.text
    if output_json:
        try:
            body = _json_mod.dumps(resp.json(), ensure_ascii=False, indent=2)
        except _json_mod.JSONDecodeError:
            pass

    click.echo(body)
